addons/hr_expense/models/hr_expense.py

```diff
@@ -312,3 +286,130 @@ def action_get_attachment_view(self):
+    employee_id = fields.Many2one('hr.employee', string="Employee", required=True, readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env['hr.employee'].search([('user_id', '=', self.env.uid)], limit=1))
+    payment_status = fields.Boolean(string='Payment received by employee', help='The employee greens this button when the payment has been received.')
+    user_valid = fields.Many2one('res.users', 'Validation By', readonly=True, copy=False, states={'draft':[('readonly',False)], 'submit':[('readonly',False)]})
+    total_amount = fields.Float(string='Total Amount', store=True, compute='_compute_amount', digits=dp.get_precision('Account'))
+    company_id = fields.Many2one('res.company', string='Company', readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env.user.company_id)
+    currency_id = fields.Many2one('res.currency', string='Currency', readonly=True, states={'draft': [('readonly', False)]}, default=lambda self: self.env.user.company_id.currency_id)
+    attachment_number = fields.Integer(compute='_compute_attachment_number', string='Number of Attachments')
+
+    @api.multi
+    def write(self, vals):
+        old_line_ids = set(self.expense_line_ids.ids)
+        res = super(HrExpenseSheet, self).write(vals)
+        if vals.get('expense_line_ids'):
+            new_line_ids = set(self.expense_line_ids.ids)
+            rec_to_update = list(old_line_ids - new_line_ids)
+            self.env['hr.expense'].browse(rec_to_update).write({'sheet_id': None, 'state': 'draft'})
+        return res
+
+    @api.multi
+    def _track_subtype(self, init_values):
+        self.ensure_one()
+        if 'state' in init_values and self.state == 'approve':
+            return 'hr_expense.mt_expense_approved'
+        elif 'state' in init_values and self.state == 'submit':
+            return 'hr_expense.mt_expense_confirmed'
+        elif 'state' in init_values and self.state == 'cancel':
+            return 'hr_expense.mt_expense_refused'
+        return super(HrExpenseSheet, self)._track_subtype(init_values)
+
+    @api.depends('expense_line_ids')
+    def _compute_amount(self):
+        for sheet in self:
+            total = 0.0
+            for line in sheet.expense_line_ids:
+                total += line.total_amount
+            sheet.total_amount = total
+
+    @api.multi
+    def _compute_attachment_number(self):
+        for sheet in self:
+            total = 0.0
+            for line in sheet.expense_line_ids:
+                total += line.attachment_number
+            sheet.attachment_number = total
+
+    @api.multi
+    def refuse_expenses(self, reason):
+        self.write({'state': 'cancel'})
+        if self.employee_id.user_id:
+            body = (_("Your Expense %s has been refused.<br/><ul class=o_timeline_tracking_value_list><li>Reason<span> : </span><span class=o_timeline_tracking_value>%s</span></li></ul>") % (self.name, reason))
```

Comment:
From now on, we try to put html inside a qweb template. Then just call message_post_view_view (I think, to check in mail_thread).